File: openacc/cublas_c2/acc.c
#include <stdio.h>
#include <stdlib.h>
#include <cublas_v2.h>
#include <sys/time.h>

int main(int argc, char **argv)
{
  int i, j, k;
  int N;
  float *A, *B, *C;
  float alpha=1.0f, beta=1.0f;
  cublasHandle_t handle;
  struct timeval tv1, tv2;

  if(argc!=2){
	printf("usage: %s N\n", argv[0]);
	return -1;
  }

  N = atoi(argv[1]);
  printf("N = %d\n", N);

  A = (float*)malloc(sizeof(float)*N*N);
  B = (float*)malloc(sizeof(float)*N*N);
  C = (float*)malloc(sizeof(float)*N*N);
  for(i=0;i<N;i++){
	for(j=0;j<N;j++){
	  C[i*N+j] = 0.0f;
	  A[i*N+j] = (float)(i) + (float)j/10.0f;
	  B[i*N+j] = (float)(i) + (float)j/10.0f;
	}
  }

#pragma acc enter data copyin(A[0:N*N], B[0:N*N], C[0:N*N])
  gettimeofday(&tv1,NULL);
#pragma acc kernels
#pragma acc loop independent collapse(2)
  for(i=0; i<N; i++){
	for(j=0; j<N; j++){
#pragma acc loop seq
	  for(k=0; k<N; k++){
		C[i*N+j] += A[i*N+k] * B[k*N+j];
	  }
	}
  }
  gettimeofday(&tv2,NULL);
#pragma acc exit data copyout(C[0:N*N])

  //t2 = omp_get_wtime();
  //printf("TIME: %d %f\n", N, t2-t1);
  printf("TIME: %d %f\n", N, (tv2.tv_sec+tv2.tv_usec*1e-6)-(tv1.tv_sec+tv1.tv_usec*1e-6));

  free(A); free(B); free(C);
  return 0;
}

File: openacc/cublas_f2/acc.f90
! -*- f90 -*-

program main
  use cublas
  use cudafor
  implicit none

  real(4), allocatable, dimension(:,:) :: A, B, C
  real(4) :: alpha, beta
  integer :: N, I, J, K
  integer :: t1, t2, t_rate, t_max, diff
  integer(kind=cuda_stream_kind) :: stream

  character*100 TMPSTR
  INTEGER iargc, nargc
  EXTERNAL iargc
  nargc=iargc()
  if(nargc.eq.0)then
     write(*,*)"usage: ./execfile N"
     stop
  else
     call getarg(1,TMPSTR)
     read(TMPSTR,*)N
     write(*,*)"N=",N
  endif

  allocate(A(N,N), B(N,N), C(N,N))

  C = 0.0
  do I=1, N
     do J=1, N
        A(J,I) = real(I-1) + real(J-1)/10.0
        B(J,I) = real(I-1) + real(J-1)/10.0
     enddo
  enddo
  alpha = 1.0;  beta = 1.0

!$acc enter data copyin(A(1:N,1:N), B(1:N,1:N), C(1:N,1:N))
  call system_clock(t1)
!$acc kernels
!$acc loop collapse(2)
  do I=1, N
     do J=1, N
!$acc loop seq
        do K=1, N
           C(J,I) = C(J,I) + A(K,I) * B(J,K)
        enddo
     enddo
  enddo
!$acc end kernels
  call system_clock(t2, t_rate, t_max)
!$acc exit data copyout(C(1:N,1:N))

  if(t2.lt.t1)then
     diff = (t_max - t1) + t2 + 1
  else
     diff = t2 - t1
  endif
  print "(A, F20.8)", "TIME:", diff/dble(t_rate)

  deallocate(A, B, C)
end program main
